pkg/ipamd/datastore/data_store.go

```diff
@@ -166,15 +161,119 @@ func prometheusRegister() {
 }
 
 // NewDataStore returns DataStore structure
-func NewDataStore(log logger.Logger) *DataStore {
+func NewDataStore(log logger.Logger, backingStore string) *DataStore {
 	prometheusRegister()
 	return &DataStore{
-		eniIPPools: make(map[string]*ENIIPPool),
-		podsIP:     make(map[PodKey]PodIPInfo),
-		log:        log,
+		eniIPPools:   make(map[string]*ENIIPPool),
+		podsIP:       make(map[string]PodIPInfo),
+		log:          log,
+		backingStore: backingStore,
 	}
 }
 
+const storageVersion = "vpc-cni-ipam/1"
+
+type storage struct {
+	Version     string         `json:"version"`
+	Allocations []storageEntry `json:"allocations"`
+}
+
+// A "row" in the conceptual IPAM datastore
+type storageEntry struct {
+	SandboxID string `json:"sandboxID"`
+	IPv4      string `json:"ipv4"`
+}
+
+// ReadBackingStore initialises the IP allocation state from the
+// configured backing store.  Should be called before using data
+// store.
+func (ds *DataStore) ReadBackingStore() error {
+	ds.log.Infof("Reading ipam state from %s", ds.backingStore)
+
+	f, err := os.Open(ds.backingStore)
+	if os.IsNotExist(err) {
+		// Ok, and a no-op.
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("datastore: error reading backing store: %v", err)
+	}
+	defer f.Close()
+
+	var data storage
+	if err := json.NewDecoder(f).Decode(&data); err != nil {
+		return fmt.Errorf("datastore: backing store is corrupted (%v) - may require recreating local pods (reboot this node)", err)
+	}
+
+	if data.Version != storageVersion {
+		return fmt.Errorf("datastore: unknown backing store format (%s != !s) - wrong CNI/ipamd version? (Rebooting this node will restart local pods and probably help)", data.Version, storageVersion)
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	eniIPs := make(map[string]*ENIIPPool)
+	for _, eni := range ds.eniIPPools {
+		for _, addr := range eni.IPv4Addresses {
+			eniIPs[addr.Address] = eni
+		}
+	}
+
+	for _, allocation := range data.Allocations {
+		eni := eniIPs[allocation.IPv4]
+		if eni == nil {
+			ds.log.Infof("datastore: Sandbox %s uses unknown IPv4 %s - presuming stale/dead", allocation.SandboxID, allocation.IPv4)
+			continue
+		}
+
+		addr := eni.IPv4Addresses[allocation.IPv4]
+
+		ds.assignPodIPv4AddressUnsafe(allocation.SandboxID, eni, addr)
+	}
+
+	ds.log.Debugf("Completed reading ipam state from %s", ds.backingStore)
+	return nil
+}
+
+func (ds *DataStore) writeBackingStoreUnsafe() error {
```

Comment:
When will this be called, and do we need to lock the datastore until this is persisted? What happens if we generate the `data` struct, but before it's serialized and flushed to disk we free or use one IP? Or if we get an error back for some reason, what should we do If the aws-node pod restarts to rebuild the sandbox state?

Reply:
Not glued in yet.  I'm going to call it while holding the datastore lock, during every new IP allocation.